### PR TITLE
SQL Connector initialization changes

### PR DESCRIPTION
**Pull request by anupama-pathirage**
With this PR we can initialize the SQL Connector for the known sql databases easily without knowing jdbc related details explicitly.   

**Examples:**

- Initialize connector for Mysql DB with default connection pool settings

```java
sql:ConnectionProperties props = {}; //Empty struct is defined, so it use default properties.
sql:ClientConnector testDB = create sql:ClientConnector(sql:MYSQL, "localhost", 3306, "testdb", "root", "root", props);
```

- Initialize connector for Mysql DB by overriding some default connection pool configurations.

```java
sql:ConnectionProperties props = {maximumPoolSize:1, poolName:"testPool"}; //Overriding required pool properties.
sql:ClientConnector testDB = create sql:ClientConnector(sql:MYSQL, "localhost", 3306, "testdb", "root", "root", props);
```

For more advanced usages such as to initialize the connector for a Database which does not have first class support or to use a more customized properties with a known DB we can use below two methods.

- Initialize connector for any DB by providing the DB specific properties by directly using the underline driver implementation class

```java
sql:ConnectionProperties props = {dataSourceClassName:"com.mysql.jdbc.jdbc2.optional.MysqlDataSource",
          datasourceProperties:{"url":"jdbc:mysql://localhost:3306/testdb","initialTimeout":100, "maxRows" : 1000}};
sql:ClientConnector testDB = create sql:ClientConnector("", "", 0, "", "root", "root", props);
```

- Initialize connector for any new DB type by providing the jdbc url directly. We can use this in the same way to have a more customized url for the known db types as well.

```java
sql:ConnectionProperties Properties = {url:"jdbc:dasjdriver://localhost:3306/testdb"};
sql:ClientConnector testDB = create sql:ClientConnector("", "", 0, "", "sa", "test", Properties);
```

### CODE REVIEW

File: modules/ballerina-native/src/test/resources/samples/datatableTest.bal [hunk 1]
```diff
@@ -55,10 +55,9 @@ struct ResultPrimitiveInt {
 
 
 function testGetPrimitiveTypes () (int i, int l, float f, float d, boolean b, string s) {
-    map propertiesMap = {"jdbcUrl":"jdbc:hsqldb:file:./target/tempdb/TEST_DATA_TABLE_DB",
-                            "username":"SA", "password":"", "maximumPoolSize":1};
-    sql:ClientConnector testDB = create sql:ClientConnector(propertiesMap);
-
+    sql:ConnectionProperties Properties = {maximumPoolSize:1};
```
Comment (lafernando) [inline modules/ballerina-native/src/test/resources/samples/datatableTest.bal]: variable names should start with a lowercase letter.

File: modules/ballerina-native/src/test/resources/samples/datatableTest.bal [hunk 2]
```diff
@@ -374,9 +370,9 @@ function testColumnAlias () (int i, int l, float f, float d, boolean b, string s
 }
```
Comment (lafernando) [inline modules/ballerina-native/src/test/resources/samples/datatableTest.bal]: Put a test to include the negative case of not finding a specific database type. We can pass in a invalid database type and put a JDBC URL to point to HSQLDB.

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/actions/data/sql/SQLDatasource.java [hunk 3]
```diff
@@ -54,8 +54,9 @@ public String getDatabaseName() {
 
     public SQLDatasource() {}
 
-    public boolean init(BMap options) {
-        buildDataSource(options);
+    public boolean init(BStruct options, String dbType, String hostOrPath, int port, String userName, String password,
```
Comment (anupama-pathirage) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/actions/data/sql/SQLDatasource.java]: Fixed all the places with 092974be1d10390da7809f482804b5cb011020a2 .Thanks.

File: modules/ballerina-native/src/main/ballerina/ballerina/data/sql/natives.bal [hunk 4]
```diff
@@ -2,7 +2,63 @@ package ballerina.data.sql;
...
+@doc:Field {value:"url: Platform independent DB access URL"}
+@doc:Field {value:"dataSourceClassName: Name of the DataSource class provided by the JDBC driver"}
+@doc:Field {value:"userName: Default authentication username used when obtaining Connections from the underlying driver"}
+@doc:Field {value:"password: Default authentication password used when obtaining Connections from the underlying driver"}
+@doc:Field {value:"connectionTestQuery: Query that will be executed to validate that the connection to the database is still alive"}
+@doc:Field {value:"poolName: User-defined name for the connection pool and appears mainly in logging"}
+@doc:Field {value:"catalog: Catalog of connections created by this pool"}
+@doc:Field {value:"connectionInitSql:  SQL statement that will be executed after every new connection creation before adding it to the pool"}
+@doc:Field {value:"driverClassName: Fully qualified Java class name of the JDBC driver to be used"}
+@doc:Field {value:"transactionIsolation:  Transaction isolation level of connections returned from the pool"}
```
Comment (anupama-pathirage) [inline modules/ballerina-native/src/main/ballerina/ballerina/data/sql/natives.bal]: Fixed with 092974b.Thanks.

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/actions/data/sql/SQLDatasource.java [hunk 5]
```diff
@@ -98,120 +99,100 @@ public void closeConnectionPool() {
...
+            if (!dataSourceClassName.isEmpty()) {
+                config.setDataSourceClassName(dataSourceClassName);
+            } else {
+                if (jdbcurl.isEmpty()) {
+                    jdbcurl = constructJDBCURL(dbType, hostOrPath, port, dbName, userName, password);
+                }
+                if (!jdbcurl.isEmpty()) {
+                    config.setJdbcUrl(jdbcurl);
+                } else {
+                    throw new BallerinaException("invalid db connection properties");
```
Comment (anupama-pathirage) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/actions/data/sql/SQLDatasource.java]: Fixed with 092974b.Thanks.

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/actions/data/sql/SQLDatasource.java [hunk 6]
```diff
@@ -223,20 +204,94 @@ private void buildDataSource(BMap options) {
...
-                } else if (value instanceof BBoolean) {
-                    config.addDataSourceProperty(keyValue, Boolean.parseBoolean(value.stringValue()));
-                }
+            BValue value = options.get(key);
+            if (value instanceof BString) {
+                config.addDataSourceProperty(key, value.stringValue());
+            } else if (value instanceof BInteger) {
+                config.addDataSourceProperty(key, ((BInteger) value).intValue());
+            } else if (value instanceof BBoolean) {
+                config.addDataSourceProperty(key, Boolean.parseBoolean(value.stringValue()));
```
Comment (anupama-pathirage) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/actions/data/sql/SQLDatasource.java]: Fixed with 092974b.Thanks.

File: modules/ballerina-native/src/test/resources/samples/datatableTest.bal (hunk 2 above)
Comment (anupama-pathirage) [inline modules/ballerina-native/src/test/resources/samples/datatableTest.bal]: Added a test case to SQLConnectorTest.bal with 092974b.Thanks.

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/actions/data/sql/SQLDatasource.java [hunk 7]
```diff
@@ -223,20 +204,94 @@ private void buildDataSource(BMap options) {
...
-            if (keyName.startsWith(Constants.PoolProperties.DATASOURCE)) {
-                String keyValue = keyName.substring(Constants.PoolProperties.DATASOURCE.length());
-                BValue value = options.get(key);
-                if (value instanceof BString) {
-                    config.addDataSourceProperty(keyValue, value.stringValue());
-                } else if (value instanceof BInteger) {
-                    config.addDataSourceProperty(keyValue, ((BInteger) value).intValue());
-                } else if (value instanceof BBoolean) {
-                    config.addDataSourceProperty(keyValue, Boolean.parseBoolean(value.stringValue()));
-                }
```
Comment (lafernando) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/actions/data/sql/SQLDatasource.java]: As per the offline chat, let's fix the int, float type issue in the next release. We will have to check the actual Java data types using reflection for the data source class.

Comment (anupama-pathirage) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/actions/data/sql/SQLDatasource.java]: Noted and will fix this with next release. Thanks.

**Pull request merged**